pkg/controller/machinehealthcheck/machinehealthcheck_controller.go

```diff
@@ -352,3 +377,71 @@ func isMaster(machine mapiv1.Machine, client client.Client) bool {
 	}
 	return false
 }
+
+// isMachineDisruptionAllowed returns true if the provided MachineDisruptionBudget allows any disruption
+func isMachineDisruptionAllowed(mdb *healthcheckingv1alpha1.MachineDisruptionBudget, maxDisruptedMachinSize int) bool {
+	if mdb.Status.ObservedGeneration < mdb.Generation {
+		glog.Warningf("the machine disruption budget %s is still being processed by the server", mdb.Name)
+		return false
+	}
+	if mdb.Status.MachineDisruptionsAllowed < 0 {
+		glog.Warningf("the machine disruption budget %s MachineDisruptionsAllowed is negative", mdb.Name)
+		return false
+	}
+	if len(mdb.Status.DisruptedMachines) > maxDisruptedMachinSize {
+		glog.Warningf("the machine disruption budget %s DisruptedMachines map too big - too many deletions not confirmed by MDB controller", mdb.Name)
+		return false
+	}
+	if mdb.Status.MachineDisruptionsAllowed == 0 {
+		glog.Warningf("cannot delete machine as it would violate the machine's disruption budget %s", mdb.Name)
+		return false
+	}
+
+	return true
+}
+
+func (r *ReconcileMachineHealthCheck) decrementMachineDisruptionsAllowed(machineName string, mdb *healthcheckingv1alpha1.MachineDisruptionBudget) error {
+	mdb.Status.MachineDisruptionsAllowed--
+	if mdb.Status.DisruptedMachines == nil {
+		mdb.Status.DisruptedMachines = make(map[string]metav1.Time)
+	}
+	// MachineHealthCheck controller needs to inform the MDB controller that it is about to remediate a machine
+	// so it should not consider it as available in calculations when updating MachineDisruptions allowed.
+	// If the machine is not remediated within a reasonable time limit MDB controller will assume that it won't
+	// be remediated at all and remove it from DisruptedMachines map.
+	mdb.Status.DisruptedMachines[machineName] = metav1.Time{Time: time.Now()}
+	return r.client.Update(context.TODO(), mdb)
+}
+
+// retryCheckAndDecrement tries to run checkAndDecrement command and if it fails because of the
+// update conflict error it will retry
+func (r *ReconcileMachineHealthCheck) retryCheckAndDecrement(machine *mapiv1.Machine) error {
+	var mdb *healthcheckingv1alpha1.MachineDisruptionBudget
+	err := retry.RetryOnConflict(updateMDBRetry, func() error {
+		mdbs, err := machineutil.GetMachineMachineDisruptionBudgets(r.client, machine)
+		if err != nil {
+			return err
+		}
+
+		if len(mdbs) > 1 {
+			return fmt.Errorf("machine %q has more than one MachineDisruptionBudget, which is not supported", machine.Name)
+		}
+
+		if len(mdbs) == 1 {
+			mdb = mdbs[0]
+
+			if !isMachineDisruptionAllowed(mdb, MaxDisruptedMachinSize) {
+				return fmt.Errorf("mahcine disruption is not allowed")
+			}
+
+			return r.decrementMachineDisruptionsAllowed(machine.Name, mdb)
+		}
+		return nil
+	})
+
+	if err == wait.ErrWaitTimeout {
+		err = fmt.Errorf("couldn't update MachineDisruptionBudget %q due to conflicts", mdb.Name)
```

Comment:
```suggestion
		return fmt.Errorf("Couldn't update MachineDisruptionBudget %q due to conflicts", mdb.Name)
```

Reply:
we want to update an error message just in cases when, `err == wait.ErrWaitTimeout`, but it also possible that we have other errors